astropy/utils/xml/tests/test_iterparse.py

```diff
@@ -0,0 +1,137 @@
+VOTABLE_XML = HEADER + 125*ROW + FOOTER
+
+# UngzipFileWrapper() behaves wraps an existing file-like Object,
+# decompressing the content content before returning it.
+# This therefore emulated the behaviour of the Python 'requests'
+# library when transparently decompressing Gzip HTTP responses.
+#
+# The critical behaviour is that---because of the
+# decompression---read() can return considerably more
+# bytes than were requested!  (But, can also return less).
+#
+# inspiration: http://stackoverflow.com/questions/4013843/how-to-wrap-file-object-read-and-write-operation-which-are-readonly
+class UngzipFileWrapper(object):
+    def __init__(self, fd, **kwargs):
+        self._file = fd
+        self._z = zlib.decompressobj(16 + zlib.MAX_WBITS)
+    def read(self, requested_length):
+        # emulate network buffering dynamics by clamping the read size
+        clamped_length = max(1, min(1<<24, requested_length))
+        compressed = self._file.read(clamped_length)
+        plaintext = self._z.decompress(compressed)
+        print ('read(len(requested)={:d},len(clamped)={:d},len(compressed)={:d},len(plaintext)={:d})'
+               .format(requested_length, clamped_length, len(compressed), len(plaintext)))
+        # Only for real local files---just for the testcase
+        if len(compressed) == 0:
+            self.close()
+        return plaintext
+    def __getattr__(self, attr):
+        return getattr(self._file, attr)
+
+# test_iterparser_over_read_simple() is a very cut down test,
+# of the original more flexible test-case, but without external
+# dependencies.  The plaintext is compressed and then decompressed
+# to provide a better emulation of the original situation where
+# the bug was observed.
+#
+# If a dependency upon 'zlib' is not desired, it would be possible to
+# simply this testcase by replacing the compress/decompression with a
+# read() method emulation that always returned more from a buffer that
+# was requested.
+def test_iterparser_over_read_simple():
+    # Take the plaintext of 512 tags, and compression it with a
+    # Gzip-style header (+16), to most closely emulate the behaviour
+    # of most HTTP servers.
+    zlib_GZIP_STYLE_HEADER = 16
+    compo = zlib.compressobj(zlib.Z_BEST_COMPRESSION,
+                             zlib.DEFLATED,
+                             zlib.MAX_WBITS + zlib_GZIP_STYLE_HEADER)
+
+    # Bytes vs. String  .encode()/.decode() for compatibility with Python 3.5.
+    s = compo.compress(VOTABLE_XML.encode())
+    s = s + compo.flush()
+    fd = BytesIO(s)
+    fd.seek(0)
+
+    # Finally setup the test of the C-based '_fast_iterparse()' iterator
+    # and a situation in which it can be called by the VOTable Parser.
+    MINIMUM_REQUESTABLE_BUFFER_SIZE = 1024
+    uncompressed_fd = UngzipFileWrapper(fd)
+    iterable = _fast_iterparse(uncompressed_fd.read, MINIMUM_REQUESTABLE_BUFFER_SIZE)
+    votf = tree.VOTableFile()
+    votf.parse(iterable, config={})
+
+if __name__ == '__main__':
```

Comment:
@pllim: for Line 135: yes, happy to remove the `if __name__ == '__main__'`.  For Lines 132‒133, these trigger the iterator in the circumstances that cause the bug to be demonstrated (failing before) as fixed by the proposed patch.  _Feedback welcomed for improved the comment the wording of the comment to make this clearer._

(Will try and see if a simpler way can be found—even if not so mirroing the original bug as closely; this should be possible because the triggering happens to be the resulting of processing too much in a single call, so it should be possible to call and ignore the results).